python/thunder/rdds/series.py

```diff
@@ -636,6 +636,111 @@ def query(self, inds, var='inds', order='F', onebased=True):
+        sorted and no records have been removed by filtering, then the resulting output should be equivalent
+        to what one would get from calling myImages.saveAsBinarySeries().
+
+        The json config file written out by this method takes the value of the 'nvalues' key from the
+        length of this object's .index attribute. If the index has been modified to have a different length
+        than the number of values in a Series record, then the resulting data files will be read back in
+        incorrectly.
+
+        Parameters
+        ----------
+        outputdirname : string path or URI to directory to be created
+            Output files will be written underneath outputdirname. This directory must not yet exist
+            (unless overwrite is True), and must be no more than one level beneath an existing directory.
+            It will be created as a result of this call.
+
+        overwrite : bool
+            If true, outputdirname and all its contents will be deleted and recreated as part
+            of this call.
+        """
+        import cStringIO as StringIO
+        import struct
+        from thunder.rdds.imgblocks.blocks import SimpleBlocks
+        from thunder.rdds.fileio.writers import getParallelWriterForPath
+        from thunder.rdds.fileio.seriesloader import writeSeriesConfig
+
+        def partitionToBinarySeries(kvIter):
+            """Collects all Series records in a partition into a single binary series record.
+            """
+            keypacker = None
+            firstKey = None
+            buf = StringIO.StringIO()
+            for seriesKey, series in kvIter:
+                if keypacker is None:
+                    keypacker = struct.Struct('h'*len(seriesKey))
+                    firstKey = seriesKey
+                # print >> sys.stderr, seriesKey, series, series.tostring().encode('hex')
+                buf.write(keypacker.pack(*seriesKey))
+                buf.write(series.tostring())
+            val = buf.getvalue()
+            buf.close()
+            # we might have an empty partition, in which case firstKey will still be None
+            if firstKey is None:
+                return iter([])
+            else:
+                label = SimpleBlocks.getBinarySeriesNameForKey(firstKey) + ".bin"
+                return iter([(label, val)])
+
+        writer = getParallelWriterForPath(outputdirname)(outputdirname, overwrite=overwrite)
+
+        binseriesrdd = self.rdd.mapPartitions(partitionToBinarySeries)
+
+        binseriesrdd.foreach(writer.writerFcn)
+        # note that we use len(self.index) here for nvalues! we don't really have a great alternative
+        # to this at the moment, short of calling first(). 
+        writeSeriesConfig(outputdirname, len(self.dims), len(self.index), dims=self.dims.count,
+                          keytype='int16', valuetype=self.dtype, overwrite=overwrite)
```

Comment:
Thoughts. Yes. So despite my rant above about the Series binary format becoming unexpectedly complex, I don't see forcing keys and values to be the same data type to be a useful simplification (that's why I changed it in the first place :) ). 

*Writing* one of these files from matlab is still simple-ish - you concatenate your keys and your values, write it out as a gigantic array, and, oh, oops, write out a JSON file specifying that both keys and values are float64. (I'm not aware of an easy way to read or write JSON from matlab? From python of course it's trivial.)

So the issue comes in if someone wanted to use matlab to open up a Series data file that we produced. In return for an extra 144 bytes / record (assuming 3d data saved as float64) and the aesthetic ugliness of representing integer indices as floats, we gain the ability to slurp up this data as a single `read` call or whatever, after which we still have to split the array into keys and values according to an `nkeys` parameter read from JSON. In contrast, as it currently stands, one would have to read this data either through a bunch of `fread` calls specifying the appropriate data type, or possibly still by slurping it as a byte array or something and doing some clever casting afterwards. In numpy this can be read pretty easily with the `struct` module, which is how we're reading and writing it ourselves.

I guess I really don't see this as a useful or cost-effective simplification. Series data is I think well past the point where we can hope to represent it as a single flat array - it's got keys, values, metadata, all kinds of stuff. If we want people to be able to easily read and write it from matlab or numpy, possibly we should be providing our own file i/o routines to be used. And if we're doing that, we should potentially also revisit the format as a whole, as I suggested above, and see if we can't find some more principled and general way to represent all of this.